extensions/wikia/CategoryPage3/CategoryPage3Hooks.class.php

```diff
@@ -0,0 +1,222 @@
+
+		foreach ( array_keys( $categories ) as $categoryTitle ) {
+			$title = Title::newFromText( $categoryTitle, NS_CATEGORY );
+
+			CategoryPage3CacheHelper::setTouched( $title );
+		}
+
+		return true;
+	}
+
+	/**
+	 * @param Title $title
+	 * @param Article $article
+	 * @return bool
+	 */
+	public static function onArticleFromTitle( $title, &$article ): bool {
+		if ( $title->isRedirect() ) {
+			$title = $article->getRedirectTarget();
+		}
+
+		if ( is_null( $title ) || !$title->inNamespace( NS_CATEGORY ) ) {
+			return true;
+		}
+
+		$layout = static::getLayout( $title, $article );
+
+		switch ( $layout ) {
+			case CategoryPageWithLayoutSelector::LAYOUT_MEDIAWIKI:
+				$article = new CategoryPageMediawiki( $title );
+				break;
+			case CategoryPageWithLayoutSelector::LAYOUT_CATEGORY_EXHIBITION:
+				$article = new CategoryExhibitionPage( $title );
+				break;
+			default:
+				$article = new CategoryPage3( $title );
+		}
+
+		return true;
+	}
+
+	/**
+	 * Redirect to the canonical category URL if any of the older
+	 * pagination URL params were used.
+	 *
+	 * @param Title $title
+	 * @param $unused
+	 * @param OutputPage $output
+	 * @param User $user
+	 * @param WebRequest $request
+	 * @param MediaWiki $mediawiki
+	 * @return bool
+	 */
+	public static function onBeforeInitialize(
+		Title $title, $unused, OutputPage $output,
+		User $user, WebRequest $request, MediaWiki $mediawiki
+	): bool {
+		$article = Article::newFromTitle( $title, $output->getContext() );
+
+		if ( $title->isRedirect() ) {
+			$title = $article->getRedirectTarget();
+		}
+
+		if (
+			is_null( $title ) ||
+			!$title->inNamespace( NS_CATEGORY ) ||
+			static::getLayout( $title, $article ) !== CategoryPageWithLayoutSelector::LAYOUT_CATEGORY_PAGE3
+		) {
+			return true;
+		}
+
+		$oldParamsUsed = array_intersect( static::$oldQueryParams, array_keys( $request->getQueryValues() ) );
+
+		if ( !empty( $oldParamsUsed ) ) {
+			$output->redirect( $title->getFullURL(), '301', 'CanonicalCategoryURL' );
+		}
+
+		return true;
+	}
+
+	public static function onGetPreferences( $user, &$defaultPreferences ) {
+		$preference = [
+			'label-message' => 'category-page3-layout-selector-label',
+			'section' => 'personal/appearance',
+			'type' => 'select'
+		];
+
+		$options = [];
+		$options[ wfMessage( 'category-page3-layout-selector-category-page3' )->escaped() ] = CategoryPageWithLayoutSelector::LAYOUT_CATEGORY_PAGE3;
+		$options[ wfMessage( 'category-page3-layout-selector-mediawiki' )->escaped() ] = CategoryPageWithLayoutSelector::LAYOUT_MEDIAWIKI;
+		$options[ wfMessage( 'category-page3-layout-selector-category-exhibition' )->escaped() ] = CategoryPageWithLayoutSelector::LAYOUT_CATEGORY_EXHIBITION;
+		$preference['options'] = $options;
+
+		$defaultPreferences[ static::GLOBAL_PREFERENCE_NAME ] = $preference;
+
+		return true;
+	}
+
+	public static function onLinkerMakeExternalLink( string $url, string $text, string &$link, array $attribs ): bool {
```

Comment:
seems like there's some logic in here, do you think it'd be worth to add a unit test?